daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -415,163 +415,156 @@ object ScenarioLedger {
+            activeContracts = prevState.activeContracts,
+            activeKeys = prevState.activeKeys,
+          )
+        }
+        processNodes(cache1, restENPs)
+      }
+      case (processingNode @ ProcessingNode(
+            mbParentId,
+            mbRollbackAncestorId,
+            nodeId :: restOfNodeIds,
+            optPrevState,
+          )) :: restENPs =>
+        val eventId = EventId(trId.id, nodeId)
+        richTr.transaction.nodes.get(nodeId) match {
+          case None =>
+            crash(s"processTransaction: non-existent node '$eventId'.")
+          case Some(node) =>
+            val newLedgerNodeInfo = LedgerNodeInfo(
+              node = node,
+              optLocation = locationInfo.get(nodeId),
+              transaction = trId,
+              effectiveAt = richTr.effectiveAt,
+              disclosures = Map.empty,
+              referencedBy = Set.empty,
+              consumedBy = None,
+              rolledbackBy = mbRollbackAncestorId,
+              parent = mbParentId.map(EventId(trId.id, _)),
+            )
+            val newCache =
+              cache0.copy(nodeInfos = cache0.nodeInfos + (eventId -> newLedgerNodeInfo))
+            val idsToProcess = processingNode.copy(children = restOfNodeIds) :: restENPs
+
+            node match {
+              case rollback: Node.Rollback =>
+                val rollbackState =
+                  RollbackBeginState(newCache.activeContracts, newCache.activeKeys)
+                processNodes(
+                  newCache,
+                  ProcessingNode(
+                    Some(nodeId),
+                    Some(nodeId),
+                    rollback.children.toList,
+                    Some(rollbackState),
+                  ) :: idsToProcess,
                 )
-              }
-              processNodes(Right(cache1), restENPs)
-            }
-            case (processingNode @ ProcessingNode(
-                  mbParentId,
-                  mbRollbackAncestorId,
-                  nodeId :: restOfNodeIds,
-                  optPrevState,
-                )) :: restENPs =>
-              val eventId = EventId(trId.id, nodeId)
-              richTr.transaction.nodes.get(nodeId) match {
-                case None =>
-                  crash(s"processTransaction: non-existent node '$eventId'.")
-                case Some(node) =>
-                  val newLedgerNodeInfo = LedgerNodeInfo(
-                    node = node,
-                    optLocation = locationInfo.get(nodeId),
-                    transaction = trId,
-                    effectiveAt = richTr.effectiveAt,
-                    disclosures = Map.empty,
-                    referencedBy = Set.empty,
-                    consumedBy = None,
-                    rolledbackBy = mbRollbackAncestorId,
-                    parent = mbParentId.map(EventId(trId.id, _)),
+
+              case nc: Node.Create =>
+                val newCache1 = newCache.createdIn(nc.coid, eventId)
```

Comment:
It looks like `markAsActive` is now unused and I believe the same holds for `addKey` and `removeKey`. Can we delete them?

Reply:
It also transpired that `nodeInfoByCoid` was also unused - hence deleted